### PR TITLE
EZP-26700: Provide better theming support to allow overwriting content editing form rows

### PR DESCRIPTION
**Pull request by webhdx**
> JIRA: https://jira.ez.no/browse/EZP-26700

# Description
This PR improves theming for content editing form. It allows to overwrite whole rows which is a ground base for some fieldtypes with specific markup i.e. `ezauthor`.

# TODO
- [x] fix behat tests

### CODE REVIEW

**Comment by webhdx** [comment]
You can ignore @ezrobot's `php-cs-fixer` report. It's now fixed.

**Comment by glye** [comment]
@webhdx If it's fixed you can delete the report, to reduce clutter.

**Comment by webhdx** [comment]
DateTime issue is sorted out and will be turned into separate PR soon.

Indeed there was a problem with `ezkeyword` field type, it's fixed now.

File: lib/FieldType/Mapper/CountryFormMapper.php [hunk 1]
```diff
@@ -8,27 +8,30 @@
...
     protected $countriesInfo;
 
-    /**
-     * @param array $countriesInfo
-     */
-    public function __construct(array $countriesInfo)
+    /** @var FieldTypeService */
+    private $fieldTypeService;
+
+    public function __construct(array $countriesInfo, FieldTypeService $fieldTypeService)
```
[the remark this answers is not in the document]
Comment (andrerom) [reply lib/FieldType/Mapper/CountryFormMapper.php]: @webhdx As it is constructor it should be fine, but changes like this still needs to be documented as a BC _change_. [Like in kernel](https://github.com/ezsystems/ezpublish-kernel/blob/master/doc/bc/changes-6.9.md) that would be `doc/bc/changes-1.10.md` in repo here.

[the remark this answers is not in the document]
Comment (webhdx) [reply lib/FieldType/Mapper/CountryFormMapper.php]: @bdunogier @andrerom I don't think we can pull it off without doing some BC breaks here and there. It's possible to avoid BC breaks but the drawback would be a lot of duplicated code (duplicating FieldType's `fromHash()` functionality in many cases). This particular case (Country fieldtype) is actually my mistake as it doesn't use `FieldTypeService` at all and I will remove it in a minut. RichText and Relation (single and multiple) also had constructor changed.

Since we are following semver and BC breaks now shouldn't be an option right now, do you think we should bump `repository-forms` to 2.x to ship it with Hybrid UI?

[the remark this answers is not in the document]
Comment (andrerom) [reply lib/FieldType/Mapper/CountryFormMapper.php]: @webhdx let me rephrase my reply above: I don't think constructor changes is a BC break in semver sense, it's something we do all the time in kernel and most other repos when we need to fix something and need more dependencies to do it. So just add a bc note about it, it's not strictly needed, but an extra heads up to those who _extend_ our classes. Breaking _usage_, aka breaking api would be something else.

File: lib/FieldType/Mapper/KeywordFormMapper.php [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * FormMapper for ezkeyword FieldType.
+ */
+class KeywordFormMapper implements FieldDefinitionFormMapperInterface, FieldValueFormMapperInterface
+{
+    public function mapFieldDefinitionForm(FormInterface $fieldDefinitionForm, FieldDefinitionData $fieldDefinition)
```
[the remark this answers is not in the document]
Comment (webhdx) [reply lib/FieldType/Mapper/KeywordFormMapper.php]: Fixed.

File: lib/FieldType/Mapper/CountryFormMapper.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (webhdx) [reply lib/FieldType/Mapper/CountryFormMapper.php]: Now clear. I added a document describing BC changes in 1.10 (I suppose this is next release number). Thanks.

**Comment by webhdx** [comment]
@bdunogier It better organizes what was added in #134. To be precise - proper handling of compound form widgets. It should handle most cases automatically without needing to touch templates as long as the fieldtype is using built in Symfony types. 

Documenting it is probably going to be a bigger story as with those changes as well as switching to `ezsystems/repository-forms` for content editing is making FieldTypes development a little bit more complex. We should gather new best practices after finishing work on fieldtypes integration.

**Comment by webhdx** [comment]
As per @andrerom request I break this PR down into parts. This PR is now a theming related and should be merged before any other related PR.

**Comment by andrerom** [comment]
Thanks @webhdx.
@glye might be easier to have a look now and review.

File: bundle/Resources/views/Content/form_fields.html.twig [hunk 3]
```diff
@@ -1,39 +1,117 @@
...
+            {{ form_errors(form) }}
+        </div>
+        <div class="ez-{{ class_prefix }}-edit-ui">
+            {{ form_widget(form) }}
+        </div>
+    </div>
+{%- endblock form_row -%}
+
+{%- block form_label -%}
+    {# Based on default Symfony label. Removed "required" class. #}
```
Comment (dpobel) [inline bundle/Resources/views/Content/form_fields.html.twig]: is removing the `required` class the only difference ? if so I think we can live with this `required` class to avoid duplicating code

**Comment by andrerom** [comment]
@webhdx Any input on the open questions here?

File: bundle/Resources/views/Content/form_fields.html.twig (hunk 3 above)
Comment (webhdx) [reply bundle/Resources/views/Content/form_fields.html.twig]: Yes, it was the only difference. Let's leave `required` class on `<label></label>` then.

**Comment by webhdx** [comment]
@andrerom Fixed according to @dpobel's review. The last thing regarding this ticket (indirectly) is to write down how to make use of the `repository-forms` by fieldtype developers. This is a bigger task and I will need some help from @mere-teresa or @DominikaK. Anyway this is ready to be merged if none of you have more feedback.

**Comment by webhdx** [comment]
Travis fails but this isn't related to my changes. I've tested it locally.

File: bundle/Resources/views/Content/form_fields.html.twig [hunk 4]
```diff
@@ -1,39 +1,97 @@
...
-                </div>
-            </div>
-        {% endfor %}
+    <fieldset {{ block('widget_container_attributes') }}>
+        {{- block('form_rows') -}}
         {{- form_rest(form) -}}
     </fieldset>
 {%- endblock form_widget_compound -%}
 
+{# Enrichements to default Symfony types #}
```
Comment (glye) [inline bundle/Resources/views/Content/form_fields.html.twig]: Tiny nitpick: `Enrichments` sorry :)

Comment (webhdx) [reply bundle/Resources/views/Content/form_fields.html.twig]: Fixed.

**Comment by andrerom** [comment]
Great!
Last remaining step, @webhdx what's up with the test failures? :)

**Comment by webhdx** [comment]
image

I'm not sure if this is related to my PR. Can someone rerun Behat tests on Travis?

**Comment by andrerom** [comment]
From what I see there are several failures on both PHP 5.6 and 7.0:
```
    Given that I have permission to create folders          # EzSystems\RepositoryForms\Features\Context\ContentEdit::thatIHavePermissionToCreateFolders()
      Form field with id|name|label|value|placeholder "Username" not found. (Behat\Mink\Exception\ElementNotFoundException)
```

So I have restarted, but doubt it will change anything.

**Comment by mnocon** [comment]
I'll look into that

**Comment by mnocon** [comment]
Please keep in mind that it's a short investigation, but it looks to me like the failures are not related to the latest Selenium/Chrome changes: the failed scenarios:

    create_without_draft.feature
    checkbox_fieldtype_edit_form.feature
    selection_fieldtype_edit_form.feature
    textline_fieldtype_edit_form.feature
    user_fieldtype_edit_form.feature
    user_registration.feature

are not marked with @javascript tag - meaning that they will be executed with a default driver (Goutte), without using Selenium at all. Running them locally seems to confirm that, as a browser is not started.

**Comment by andrerom** [comment]
I was not really hinting about selenium change here, rather if the failure sounded familiar `Form field with ... "Username" not found`

**Comment by mnocon** [comment]
@webhdx @andrerom With help from @damianz and @alongosz I've created a PR that changes Travis configuration to the one we used before: https://github.com/ezsystems/repository-forms/pull/154

We've tested it on the master branch on Travis and with that the tests are passing again.

**Comment by andrerom** [comment]
@mnocon Thanks, merged. @webhdx can you rebase?

**Comment by webhdx** [comment]
We're back on rails, tests are passing. I believe we can merge this one and see what's next.
